cl-postgres/protocol.lisp

```diff
@@ -212,35 +212,63 @@ this raises a condition."
-                         (error 'database-error :message "Got GSS init message when a context was already established"))
-                       (init-gss-msg nil))
-                    (8 (unless gss-context
-                         (error 'database-error :message "Got GSS continuation message without a context"))
-                       (init-gss-msg (read-bytes socket (- size 4)))))))))))
-  (loop
-   (message-case socket
-     ;; ReadyForQuery
-     (#\Z (read-uint1 socket)
-          (return))))
-  socket)
+                       ;; Authentication message
+                       (#\R (let ((type (read-uint4 socket)))
+                              (ecase type
+                                (0 (return))
+                                (2 (error 'database-error :message "Unsupported Kerberos authentication requested."))
+                                (3 (unless password (error "Server requested plain-password authentication, but no password was given."))
+                                   (plain-password-message socket password)
+                                   (force-output socket))
+                                (4 (error 'database-error :message "Unsupported crypt authentication requested."))
+                                (5 (unless password (error "Server requested md5-password authentication, but no password was given."))
+                                   (md5-password-message socket password user (read-bytes socket 4))
+                                   (force-output socket))
+                                (6 (error 'database-error :message "Unsupported SCM authentication requested."))
+                                (7 (when gss-context
+                                     (error 'database-error :message "Got GSS init message when a context was already established"))
+                                   (init-gss-msg nil))
+                                (8 (unless gss-context
+                                     (error 'database-error :message "Got GSS continuation message without a context"))
+                                   (init-gss-msg (read-bytes socket (- size 4))))
+                                (9 ) ; auth_required_sspi or auth_req_sspi sspi negotiate without wrap() see postgresql source code src/libpq/pqcomm.h
+                                (10 (scram-msg-init (read-bytes socket (- size 4)) ;(read-simple-str socket)
+                                                    )) ;auth_required_sasl or auth_req_sasl see postgresql source code src/libpq/pqcomm.h scram section
+                                (11 (scram-msg-cont (read-bytes socket (- size 4)))) ;auth_sasl_continue or auth_req_sasl_cont see postgresql source code
+                                        ;src/libpq/pqcomm.h
+                                (12 (scram-msg-fin (read-bytes socket (- size 4)))) ;auth_sasl_final or auth_req_sasl_fin see postgresql source code src/libpq/pqcomm.h
+                                ))))))
```

Comment:
Is `message-case` a macro that does not contain a `&body`? This indentation is too far to the right for my taste.

Reply:
Yes, message-case is a macro that does not contain a &body. Has been this way since Marijnh wrote that code. I have not changed that. Can reduce the indent.

Reply:
But

https://github.com/marijnh/Postmodern/blob/486f3284d1691653caad9c19871e21437a8b1e93/cl-postgres/protocol.lisp#L17

Reply:
I was looking in the wrong place. Sorry. I really appreciate you looking at the code.

Reply:
No problem with that whatsoever. Huge pull reviews get complex like that and I know that it helps to have someone fresh look at the code once the big PR is working but dirty.